### PR TITLE
Support Allocation action using new interface

### PR DESCRIPTION
**Pull request by leeyun-amzn**
Signed-off-by: Annie Lee <leeyun@amazon.com>

*Issue #, if available:* N/A

*Description of changes:*
Support Allocation action using new interface.

*CheckList:*
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/index-management/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/AllocationActionParser.kt [hunk 1]
```diff
@@ -7,19 +7,58 @@ package org.opensearch.indexmanagement.indexstatemanagement.action
...
 class AllocationActionParser : ActionParser() {
     override fun fromStreamInput(sin: StreamInput): Action {
-        TODO("Not yet implemented")
+        val require = suppressWarning(sin.readMap())
+        val include = suppressWarning(sin.readMap())
+        val exclude = suppressWarning(sin.readMap())
+        val waitFor = sin.readBoolean()
+        val index = sin.readInt()
+
+        return AllocationAction(require, exclude, include, waitFor, index)
```
Comment (downsrob) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/AllocationActionParser.kt:28]: This looks to still have the reverse order of exclude, include vs include, exclude

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/AllocationActionParser.kt [hunk 2]
```diff
@@ -7,19 +7,58 @@ package org.opensearch.indexmanagement.indexstatemanagement.action
...
+        ensureExpectedToken(Token.START_OBJECT, xcp.currentToken(), xcp)
+        while (xcp.nextToken() != Token.END_OBJECT) {
+            val fieldName = xcp.currentName()
+            xcp.nextToken()
+            when (fieldName) {
+                REQUIRE -> assignObject(xcp, require)
+                INCLUDE -> assignObject(xcp, include)
+                EXCLUDE -> assignObject(xcp, exclude)
+                WAIT_FOR -> waitFor = xcp.booleanValue()
+                else -> throw IllegalArgumentException("Invalid field: [$fieldName] found in AllocationActionConfig.")
```
Comment (downsrob) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/AllocationActionParser.kt:46]: Could you remove the 'Config' bit here

File: src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/model/ActionTests.kt [hunk 3]
```diff
@@ -117,8 +117,7 @@ class ActionTests : OpenSearchTestCase() {
         roundTripAction(randomIndexPriorityActionConfig())
     }
 
-    // TODO: fixme - enable the test
-    private fun `test allocation action round trip`() {
+    fun `test allocation action round trip`() {
         roundTripAction(randomAllocationActionConfig(require = mapOf("box_type" to "hot")))
```
Comment (downsrob) [inline src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/model/ActionTests.kt:121]: The parameter order should have been caught by a good ActionTest/XContentTest but these tests aren't great. Could you harden it by adding the other parameters to the test? Just something like 
```
require = ...,
include = mapOf(randomAlphaOfLengthBetween(1, 10) to randomAlphaOfLengthBetween(1, 10)),
exclude = ...,
```

Comment (downsrob) [reply src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/model/ActionTests.kt:121]: The same for the XContentTests

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/AllocationActionParser.kt (hunk 2 above)
Comment (leeyun-amzn) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/AllocationActionParser.kt]: Removed

**Review by downsrob — DISMISSED** [review]

**Review by thalurur — APPROVED** [review]

**Review by downsrob — APPROVED** [review]

**Review by thalurur — APPROVED** [review]

**Pull request merged**